pkg/cloud/cloud_test.go

```diff
@@ -1230,6 +1230,134 @@ func TestListSnapshots(t *testing.T) {
 	}
 }
 
+func TestWaitForAttachmentState(t *testing.T) {
+	testCases := []struct {
+		name             string
+		volumeID         string
+		expectedState    string
+		expectedInstance string
+		expectedDevice   string
+		alreadyAssigned  bool
+		expectError      bool
+	}{
+		{
+			name:             "success: attached",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeAttachedState,
+			expectedInstance: "1234",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  false,
+			expectError:      false,
+		},
+		{
+			name:             "success: detached",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeDetachedState,
+			expectedInstance: "1234",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  false,
+			expectError:      false,
+		},
+		{
+			name:             "success: disk not found, assumed detached",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeDetachedState,
+			expectedInstance: "1234",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  false,
+			expectError:      false,
+		},
+		{
+			name:             "failure: disk not found, expected attached",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeAttachedState,
+			expectedInstance: "1234",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  false,
+			expectError:      true,
+		},
+		{
+			name:             "failure: unexpected device",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeAttachedState,
+			expectedInstance: "1234",
+			expectedDevice:   "/dev/xvdbb",
+			alreadyAssigned:  false,
+			expectError:      true,
+		},
+		{
+			name:             "failure: unexpected instance",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeAttachedState,
+			expectedInstance: "1235",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  false,
+			expectError:      true,
+		},
+		{
+			name:             "failure: already assigned but wrong state",
+			volumeID:         "vol-test-1234",
+			expectedState:    volumeAttachedState,
+			expectedInstance: "1235",
+			expectedDevice:   "/dev/xvdba",
+			alreadyAssigned:  true,
+			expectError:      true,
+		},
+	}
+
+	volumeAttachmentStatePollSteps = 1
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockEC2 := mocks.NewMockEC2(mockCtrl)
+			c := newCloud(mockEC2)
+
+			attachedVol := &ec2.Volume{
+				VolumeId:    aws.String(tc.volumeID),
+				Attachments: []*ec2.VolumeAttachment{{Device: aws.String("/dev/xvdba"), InstanceId: aws.String("1234"), State: aws.String("attached")}},
+			}
+
+			detachedVol := &ec2.Volume{
+				VolumeId:    aws.String(tc.volumeID),
+				Attachments: []*ec2.VolumeAttachment{{Device: aws.String("/dev/xvdba"), InstanceId: aws.String("1234"), State: aws.String("detached")}},
+			}
+
+			ctx := context.Background()
+
+			switch tc.name {
+			case "success: detached":
```

Comment:
(just a style alternative): if u want to avoid depending on test case name you could make these anonymous functions, something like this https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/driver/node_test.go#L69

Reply:
that's definitely cleaner. i'll use this style next time